File: backend/app/services/facebook.py
"""Facebook Graph API ingest for the official CFL Fanpage.

Fetches recent posts, then comments per post (including replies via
filter=stream), and persists them the same way as the CSV/Store pipelines.
Group data has no API — that stays on the CSV upload path (see csv_ingest.py).
"""
from __future__ import annotations

import hashlib
import logging
from datetime import datetime, timezone

import httpx
from sqlalchemy.orm import Session

from ..config import settings
from ..models import Comment, IngestRun, Post

log = logging.getLogger("facebook")

GRAPH_VERSION = "v19.0"
GRAPH_BASE = f"https://graph.facebook.com/{GRAPH_VERSION}"


def _dedupe_hash(source: str, external_id: str) -> str:
    return hashlib.sha256(f"fb|{source}|{external_id}".encode("utf-8")).hexdigest()


def _parse_fb_date(value: str | None) -> datetime | None:
    if not value:
        return None
    try:
        # Graph API returns e.g. "2026-07-06T10:00:00+0700"
        return datetime.strptime(value, "%Y-%m-%dT%H:%M:%S%z")
    except ValueError:
        try:
            return datetime.fromisoformat(value)
        except ValueError:
            return None


def _graph_get(client: httpx.Client, url: str, params: dict) -> dict:
    res = client.get(url, params=params, timeout=20)
    data = res.json()
    if res.status_code != 200 or "error" in data:
        err = data.get("error", {})
        code = err.get("code")
        message = err.get("message", res.text[:200])
        if code == 190:
            raise PermissionError(
                f"Facebook Access Token đã hết hạn hoặc không hợp lệ: {message}"
            )
        raise RuntimeError(f"Facebook Graph API lỗi (code={code}): {message}")
    return data


def fetch_posts(client: httpx.Client, page_id: str, access_token: str,
                since: str | None = None, until: str | None = None,
                limit: int = 25) -> list[dict]:
    """Fetch posts with pagination, following paging.next until exhausted or empty page."""
    url = f"{GRAPH_BASE}/{page_id}/posts"
    params = {
        "access_token": access_token,
        "fields": "id,message,created_time,permalink_url",
        "limit": limit,
    }
    if since:
        params["since"] = since
    if until:
        params["until"] = until

    posts: list[dict] = []
    next_url, next_params = url, params
    while next_url:
        data = _graph_get(client, next_url, next_params)
        posts.extend(data.get("data", []))
        paging = data.get("paging", {})
        next_url = paging.get("next")
        next_params = {}  # `next` is a full URL already carrying all params
        if not data.get("data"):
            break
    return posts


def fetch_comments(client: httpx.Client, post_id: str, access_token: str,
                   limit: int = 100) -> list[dict]:
    """Fetch comments (including replies) for a single post, with pagination."""
    url = f"{GRAPH_BASE}/{post_id}/comments"
    params = {
        "access_token": access_token,
        "fields": "id,message,created_time,from,like_count",
        "filter": "stream",  # flattens top-level comments + replies
        "limit": limit,
    }
    comments: list[dict] = []
    next_url, next_params = url, params
    while next_url:
        data = _graph_get(client, next_url, next_params)
        comments.extend(data.get("data", []))
        paging = data.get("paging", {})
        next_url = paging.get("next")
        next_params = {}
        if not data.get("data"):
            break
    return comments


def ingest_facebook(db: Session, *, since: str | None = None, until: str | None = None,
                    post_limit: int = 25) -> IngestRun:
    """Pull recent Fanpage posts + comments via Graph API and persist new rows."""
    run = IngestRun(source_type="fb_page", status="running")
    db.add(run)
    db.flush()

    page_id = settings.fb_page_id
    token = settings.fb_access_token
    if not page_id or not token:
        run.status = "failed"
        run.error = "Thiếu FB_PAGE_ID hoặc FB_ACCESS_TOKEN trong .env"
        run.finished_at = datetime.now(timezone.utc)
        db.commit()
        db.refresh(run)
        return run

    fetched, new_count = 0, 0
    try:
        with httpx.Client() as client:
            posts = fetch_posts(client, page_id, token, since, until, post_limit)
            for p in posts:
                post_dedupe_key = p["id"]
                post = db.query(Post).filter_by(
                    source_type="fb_page", external_id=post_dedupe_key
                ).first()
                if post is None:
                    post = Post(
                        source_type="fb_page",
                        external_id=post_dedupe_key,
                        published_at=_parse_fb_date(p.get("created_time")),
                        message=p.get("message", ""),
                        permalink=p.get("permalink_url"),
                    )
                    db.add(post)
                    db.flush()

                comments = fetch_comments(client, p["id"], token)
                fetched += len(comments)
                for c in comments:
                    dedupe = _dedupe_hash("fb_page", c["id"])
                    if db.query(Comment.id).filter_by(dedupe_hash=dedupe).first():
                        continue
                    msg = c.get("message", "") or ""
                    comment = Comment(
                        post_id=post.id,
                        source_type="fb_page",
                        external_id=c["id"],
                        created_at=_parse_fb_date(c.get("created_time")),
                        author_hint=(c.get("from") or {}).get("name"),
                        message=msg,
                        dedupe_hash=dedupe,
                        ingest_run_id=run.id,
                        skipped_analysis=len(msg.strip()) < 2,
                    )
                    db.add(comment)
                    new_count += 1
        run.status = "done"
    except PermissionError as e:
        run.status = "failed"
        run.error = str(e)
    except Exception as e:
        log.exception("facebook ingest failed")
        run.status = "failed"
        run.error = f"Lỗi không xác định: {e}"

    run.rows_fetched = fetched
    run.rows_new = new_count
    run.finished_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(run)
    return run

File: worker/src/services/facebook.ts
// Ported from backend/app/services/facebook.py — same API contract, fetch() instead of httpx.
//
// D1/Workers subrequest budget note: Workers cap subrequests per invocation
// (50 on Free, 10,000 on Paid - see Cloudflare Workers limits). Both fetch()
// calls and D1 operations count against this. Comments are collected across
// ALL posts first, then deduped/inserted via chunked batch operations (mirrors
// csvIngest.ts / sensortower.ts) instead of one D1 round-trip per comment.
// Comment pagination per post is also capped to bound fetch() calls.
import type { Env } from "../types";
import type { IngestRunRow } from "./csvIngest";

const GRAPH_VERSION = "v19.0";
const GRAPH_BASE = `https://graph.facebook.com/${GRAPH_VERSION}`;
// Conservative defaults to fit Cloudflare Workers Free plan's 50-subrequest-per-
// invocation budget (fetch() + D1 calls share this budget). With postLimit=10 and
// 2 pages/post, worst case is ~21 fetch() calls + a handful of batched D1 calls.
// On Workers Paid (10,000 subrequests/invocation, $5/mo) these can be raised a lot
// to pull more posts/history per run - see README.
const MAX_COMMENT_PAGES_PER_POST = 2; // 2 x 100 = up to 200 comments/post/run

async function dedupeHash(source: string, externalId: string): Promise<string> {
  const buf = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(`fb|${source}|${externalId}`));
  return [...new Uint8Array(buf)].map((b) => b.toString(16).padStart(2, "0")).join("");
}

function parseFbDate(value?: string | null): string | null {
  if (!value) return null;
  const d = new Date(value);
  return isNaN(d.getTime()) ? null : d.toISOString();
}

async function graphGet(url: string, params: Record<string, string>): Promise<any> {
  const full = params && Object.keys(params).length ? `${url}?${new URLSearchParams(params)}` : url;
  const res = await fetch(full, { signal: AbortSignal.timeout(20000) });
  const data: any = await res.json();
  if (!res.ok || data.error) {
    const err = data.error || {};
    if (err.code === 190) {
      throw Object.assign(new Error(`Facebook Access Token đã hết hạn hoặc không hợp lệ: ${err.message}`), { permission: true });
    }
    throw new Error(`Facebook Graph API lỗi (code=${err.code}): ${err.message || (await res.text().catch(() => ""))}`);
  }
  return data;
}

async function fetchPosts(pageId: string, token: string, since?: string, until?: string, limit = 25): Promise<any[]> {
  // `limit` is Graph API's PAGE size, not a total cap — `paging.next` keeps
  // going through the page's entire post history regardless of it. Stop as
  // soon as we have `limit` posts total, or this can burn through Workers'
  // per-invocation subrequest budget on a page with a long post history.
  const params: Record<string, string> = {
    access_token: token, fields: "id,message,created_time,permalink_url", limit: String(limit),
  };
  if (since) params.since = since;
  if (until) params.until = until;

  const posts: any[] = [];
  let nextUrl: string | null = `${GRAPH_BASE}/${pageId}/posts`;
  let nextParams: Record<string, string> = params;
  while (nextUrl && posts.length < limit) {
    const data = await graphGet(nextUrl, nextParams);
    posts.push(...(data.data || []));
    nextUrl = data.paging?.next || null;
    nextParams = {};
    if (!data.data?.length) break;
  }
  return posts.slice(0, limit);
}

async function fetchComments(postId: string, token: string, limit = 100): Promise<any[]> {
  const params: Record<string, string> = {
    access_token: token, fields: "id,message,created_time,from,like_count", filter: "stream", limit: String(limit),
  };
  const comments: any[] = [];
  let nextUrl: string | null = `${GRAPH_BASE}/${postId}/comments`;
  let nextParams: Record<string, string> = params;
  let pages = 0;
  while (nextUrl && pages < MAX_COMMENT_PAGES_PER_POST) {
    const data = await graphGet(nextUrl, nextParams);
    comments.push(...(data.data || []));
    nextUrl = data.paging?.next || null;
    nextParams = {};
    pages++;
    if (!data.data?.length) break;
  }
  return comments;
}

function chunk<T>(arr: T[], size: number): T[][] {
  const out: T[][] = [];
  for (let i = 0; i < arr.length; i += size) out.push(arr.slice(i, i + size));
  return out;
}

export async function ingestFacebook(
  env: Env, since?: string, until?: string, postLimit = 50, note?: Record<string, any>
): Promise<IngestRunRow> {
  const db = env.DB;
  const startedAt = new Date().toISOString();
  const noteText = note
    ? JSON.stringify(note)
    : (since || until ? JSON.stringify({ start_date: since || null, end_date: until || null, post_limit: postLimit }) : null);
  const runInsert = await db
    .prepare(`INSERT INTO ingest_runs (source_type, started_at, status, note) VALUES ('fb_page', ?, 'running', ?)`)
    .bind(startedAt, noteText).run();
  const runId = runInsert.meta.last_row_id as number;

  const pageId = env.FB_PAGE_ID;
  const token = env.FB_ACCESS_TOKEN;
  if (!pageId || !token) {
    const finishedAt = new Date().toISOString();
    const error = "Thiếu FB_PAGE_ID hoặc FB_ACCESS_TOKEN trong secrets";
    await db.prepare(`UPDATE ingest_runs SET status='failed', error=?, finished_at=? WHERE id=?`)
      .bind(error, finishedAt, runId).run();
    return { id: runId, source_type: "fb_page", status: "failed", started_at: startedAt, finished_at: finishedAt, rows_fetched: 0, rows_new: 0, note: noteText, error };
  }

  let fetched = 0, newCount = 0, error: string | null = null;
  try {
    const posts = await fetchPosts(pageId, token, since, until, postLimit);

    // Resolve posts: one batched lookup for existing rows, one batched insert for new ones.
    const postExternalIds = posts.map((p) => p.id);
    const postIdByExternal = new Map<string, number>();
    for (const idsChunk of chunk(postExternalIds, 90)) {
      const placeholders = idsChunk.map(() => "?").join(",");
      const res = await db
        .prepare(`SELECT id, external_id FROM posts WHERE source_type='fb_page' AND external_id IN (${placeholders})`)
        .bind(...idsChunk).all<{ id: number; external_id: string }>();
      for (const r of res.results) postIdByExternal.set(r.external_id, r.id);
    }
    const newPosts = posts.filter((p) => !postIdByExternal.has(p.id));
    for (const c of chunk(newPosts, 50)) {
      const stmts = c.map((p) =>
        db.prepare(
          `INSERT INTO posts (source_type, external_id, published_at, message, permalink) VALUES ('fb_page', ?, ?, ?, ?)`
        ).bind(p.id, parseFbDate(p.created_time), p.message || "", p.permalink_url || null)
      );
      const results = await db.batch(stmts);
      results.forEach((r, idx) => postIdByExternal.set(c[idx].id, r.meta.last_row_id as number));
    }

    // Fetch all comments across all posts before touching D1 again.
    const allComments: { postId: number; cm: any; hash: string }[] = [];
    for (const p of posts) {
      const comments = await fetchComments(p.id, token);
      fetched += comments.length;
      const postId = postIdByExternal.get(p.id)!;
      for (const cm of comments) {
        allComments.push({ postId, cm, hash: await dedupeHash("fb_page", cm.id) });
      }
    }

    // One batched dedupe check instead of a SELECT per comment.
    const existing = new Set<string>();
    for (const c of chunk(allComments, 90)) {
      const placeholders = c.map(() => "?").join(",");
      const res = await db
        .prepare(`SELECT dedupe_hash FROM comments WHERE dedupe_hash IN (${placeholders})`)
        .bind(...c.map((x) => x.hash)).all<{ dedupe_hash: string }>();
      for (const row of res.results) existing.add(row.dedupe_hash);
    }
    const fresh = allComments.filter((x) => !existing.has(x.hash));

    for (const c of chunk(fresh, 90)) {
      const stmts = c.map(({ postId, cm, hash }) => {
        const msg = cm.message || "";
        return db.prepare(
          `INSERT INTO comments (post_id, source_type, external_id, created_at, author_hint, message, dedupe_hash, ingest_run_id, skipped_analysis)
           VALUES (?, 'fb_page', ?, ?, ?, ?, ?, ?, ?)`
        ).bind(postId, cm.id, parseFbDate(cm.created_time), cm.from?.name || null, msg, hash, runId, msg.trim().length < 2 ? 1 : 0);
      });
      await db.batch(stmts);
      newCount += c.length;
    }
  } catch (e: any) {
    error = e.message || String(e);
  }

  const finishedAt = new Date().toISOString();
  const status = error ? "failed" : "done";
  await db.prepare(`UPDATE ingest_runs SET status=?, rows_fetched=?, rows_new=?, note=?, error=?, finished_at=? WHERE id=?`)
    .bind(status, fetched, newCount, noteText, error, finishedAt, runId).run();

  return { id: runId, source_type: "fb_page", status, started_at: startedAt, finished_at: finishedAt, rows_fetched: fetched, rows_new: newCount, note: noteText, error };
}
